Combine owners and nonce tables

petarjuki7 opened the pull request:
## Issue Addressed

Addresses Issue #385 

## Proposed Changes

There is now one table `owners` which has both the `fee_recipient` and `nonce` fields.

## Additional Info

Removed `NOT NULL` on `fee_recipient`, had to do it because on the `BUMP_NONCE` command we are only inserting on `owner` and `nonce`. 
Tested with `cargo test` in the `database` crate and all the tests pass.

Side comment:
As far as I can tell `cargo test` doesn't work on the `stable` branch, because in the top level `Cargo.toml` there is a `sync` feature missing in the `tokio` dependency import.

petarjuki7 committed:
Combine owners and nonce tables

dknopik commented:
Thanks for the PR! 

>Side comment:
As far as I can tell cargo test doesn't work on the stable branch, because in the top level Cargo.toml there is a sync feature missing in the tokio dependency import.

Yeah, I also noticed that the tests for the database can not be run on their own due to this. There is some other PR where I add the "sync" feature.

>Removed NOT NULL on fee_recipient, had to do it because on the BUMP_NONCE command we are only inserting on owner and nonce.

Makes sense. But I think we might need to adjust this: https://github.com/sigp/anchor/blob/e671f385dbe8f6f0fa3c4f1eec5cfba048d654d9/anchor/database/src/validator_operations.rs#L46 to also accept NULL values by reading an `Option<String>`.

I think defaulting the nonce to `0` might cause issues, as setting the fee recipient will set the nonce to 0 if there was no row before. This is different from an unset nonce, as a set nonce of 0 implies that nonce 0 was used already. I suspect this might cause issues with syncing for owners that set the recipient before registering any validators.

I will add two test cases to illustrate these issues :)

dknopik committed:
Add tests

petarjuki7 committed:
Fix nonce and fee_recipient

petarjuki7 commented:
I changed the in-memory storage a bit.
With previous in-memory storage we would assume the nonce is always there:
`nonces: HashMap<Address, u16>`

I changed nonce storage to `HashMap<Address, Option<u16>>` to properly represent:
  - `None` = owner exists but has never used a nonce
  - `Some(0)` = owner has used nonce 0
  - `Some(n)` = owner has used nonce n

Also modified all the related nonce code to use `Option<u16>` instead of `u16`
I also updated `BUMP_NONCE` SQL to use `COALESCE(nonce, -1) + 1` to handle `NULL` nonces

I'm happy changing it either way you see fit.

A simpler solution is to leave it like it was, and handle situations with NULL nonces like this:
`let nonce: u64 = row.get::<_, Option<u64>>(0)?.unwrap_or(0);`

petarjuki7 committed:
Small style fix

petarjuki7 commented on anchor/database/src/cluster_operations.rs:
```diff
@@ -166,16 +166,17 @@ impl NetworkDatabase {
                     .expect("This must exist");
-                *entry += 1;
-                nonce = *entry;
+                let new_value = entry.unwrap_or(0) + 1;
```
Here we are sure this value exists so `unwrap()`ing may be Ok. If needed I can change to a `match`.

dknopik commented on anchor/database/src/sql_operations.rs:
```diff
@@ -85,9 +85,9 @@ pub const UPDATE_BLOCK_NUMBER: &str = r#"UPDATE block SET block_number = ?1"#;
-    INSERT INTO nonce (owner, nonce) VALUES (?1, 0)
-    ON CONFLICT (owner) DO UPDATE SET nonce = nonce + 1
+    INSERT INTO owners (owner, nonce) VALUES (?1, 0)
+    ON CONFLICT (owner) DO UPDATE SET nonce = COALESCE(nonce, -1) + 1
```
Oooh, good idea, I like it :)

I think `COALESCE(nonce + 1, 0)` might be easier to understand and should also work because `NULL + 1` computes to `NULL`.

dknopik commented on anchor/database/src/lib.rs:
```diff
@@ -109,7 +109,7 @@ struct SingleState {
     clusters: HashSet<ClusterId>,
     /// Nonce of the owner account
-    nonces: HashMap<Address, u16>,
+    nonces: HashMap<Address, Option<u16>>,
```
Hm, I think it is preferable to keep this as before. My reasoning is that at this point there should be a single representation for a certain state regardless of it's representation in the SQL database - when changed this way, an owner that has not registered a validator yet can be either represented as not being in the map or as mapping to `None`.

I feel like it may be preferable to `filter_map` in `fetch_nonces` to not insert `None`s and keep this type as before.

petarjuki7 replied:
Noted, fixed.

dknopik commented on anchor/database/src/table_schema.sql:
```diff
@@ -5,12 +5,8 @@ INSERT INTO block (block_number) VALUES (0);
 
 CREATE TABLE owners (
     owner TEXT PRIMARY KEY,
```
Btw, not directly related to your changes, but I think we need to add `NOT NULL` here as SQLite [allows `NULL` in primary keys](https://sqlite.org/lang_createtable.html#the_primary_key).

petarjuki7 committed:
filter_map on fetch_nonces

petarjuki7 committed:
lint fix

diegomrsantos commented:
> I changed the in-memory storage a bit. With previous in-memory storage we would assume the nonce is always there: `nonces: HashMap<Address, u16>`
> 
> I changed nonce storage to `HashMap<Address, Option<u16>>` to properly represent:
> 
> * `None` = owner exists but has never used a nonce
> * `Some(0)` = owner has used nonce 0
> * `Some(n)` = owner has used nonce n
> 
> Also modified all the related nonce code to use `Option<u16>` instead of `u16` I also updated `BUMP_NONCE` SQL to use `COALESCE(nonce, -1) + 1` to handle `NULL` nonces
> 
> I'm happy changing it either way you see fit.
> 
> A simpler solution is to leave it like it was, and handle situations with NULL nonces like this: `let nonce: u64 = row.get::<_, Option<u64>>(0)?.unwrap_or(0);`

Can we say an unused nonce is 0?

petarjuki7 commented:
> Can we say an unused nonce is 0?

Since nonces in Ethereum start with 0 (the first transactions nonce is 0), then maybe putting it at 0 could read as "already had a first transaction as nonce 0" and not that it hasn't been set yet

dknopik commented on anchor/database/src/sql_operations.rs:
```diff
@@ -85,9 +85,9 @@ pub const UPDATE_BLOCK_NUMBER: &str = r#"UPDATE block SET block_number = ?1"#;
-    INSERT INTO nonce (owner, nonce) VALUES (?1, 0)
-    ON CONFLICT (owner) DO UPDATE SET nonce = nonce + 1
+    INSERT INTO owners (owner, nonce) VALUES (?1, 0)
+    ON CONFLICT (owner) DO UPDATE SET nonce = COALESCE(nonce + 1, 0)
```
These are different.

dknopik commented on anchor/database/src/table_schema.sql:
```diff
@@ -4,13 +4,9 @@ CREATE TABLE block (
-    nonce INTEGER DEFAULT 0
+    owner TEXT PRIMARY KEY NOT NULL,
+    fee_recipient TEXT,
+    nonce INTEGER
```
No, we need to have the NULL (not present) value

diegomrsantos commented:
> > Can we say an unused nonce is 0?
> 
> Since nonces in Ethereum start with 0 (the first transactions nonce is 0), then maybe putting it at 0 could read as "already had a first transaction as nonce 0" and not that it hasn't been set yet

Are you sure? The yellow paper defines nonce as the number of transactions sent from the address.

diegomrsantos commented:
> > > Can we say an unused nonce is 0?
> > 
> > 
> > Since nonces in Ethereum start with 0 (the first transactions nonce is 0), then maybe putting it at 0 could read as "already had a first transaction as nonce 0" and not that it hasn't been set yet
> 
> Are you sure? The yellow paper defines nonce as the number of transactions sent from the address.

It also says "An account is empty when it has no code, zero nonce and zero balance"

dknopik commented:
We are not talking about Ethereum nonces here, but nonces as used in the SSV network. 

In the database, we store the previous, not the next nonce. Therefore there is a difference between 0 (which means the last `ValidatorAdded` processed was supposed to have a nonce of 0) and NULL (which means that there was no `ValidatorAdded` message yet).

dknopik commented:
To clear up the confusion why I put " :+1: " to 
>Since nonces in Ethereum start with 0 (the first transactions nonce is 0), then maybe putting it at 0 could read as "already had a first transaction as nonce 0" and not that it hasn't been set yet

While we do not work with Ethereum nonces, but with nonces as used in SSV's `ValidatorAdded` events, I liked that comment as it highlighted that we store the nonce of the already seen event, not of the expected nonce of the next event (which is also the current event count).

I will test this change on some nodes to ensure it's working fine, then we can merge this :) Thanks @petarjuki7!

dknopik committed:
Merge branch 'unstable' into fork/petarjuki7/merge_owners_nonce_table

dknopik committed:
Fix loading clusters

dknopik commented:
I found another small issue: we needed to account for the null value when `fetch_clusters` is run during startup. I fixed it myself, hope you don't mind :)

We will merge this soon after deciding what will go into the next version :+1:

petarjuki7 commented:
> We will merge this soon after deciding what will go into the next version 👍

Great, thanks! :)

Thought of something else that might be worth changing:
currently in the `bump_and_get_nonce()` function, we update the nonce manually in memory:
https://github.com/sigp/anchor/blob/94d9d4a678f17f1b9341c7965a1aad3891f4302a/anchor/database/src/cluster_operations.rs#L164-L180

It could be changed to:
```rust
let mut stmt = tx.prepare_cached(sql_operations::GET_NONCE)?;
let nonce: u16 = stmt.query_row(params![owner.to_string()], |row| row.get(0))?;

// Update in memory with the truth from the database
self.modify_state(|state| {
    state.single_state.nonces.insert(*owner, nonce);
});
Ok(nonce)
```
So we are sure it is the same in the database as in memory. This removes any risk of the in-memory logic diverging from the SQL logic.

This passes all the tests also, let me know what do you think, I can include it in a commit.

dknopik reviewed (approved):
LGTM, thanks @petarjuki7!

dknopik commented:
@petarjuki7 

Thanks for your suggestion!

Our caching layer operates under the assumption that the cache update is implemented consistently with the actual database upgrade. Also, no other processes modify the database - which will be enforced by #358. So I think to keep the approach to DB updates consistent, we should keep the current state.

@ThreeHrSleep maybe this is a good target for fuzz testing to ensure the update functions are correct in all cases: 
- create some input events and apply them
- capture the state
- reopen database
- check if the reopened state matches
